src/v/archival/adjacent_segment_merger.cc

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "archival/adjacent_segment_merger.h"
+
+#include "archival/ntp_archiver_service.h"
+#include "archival/segment_reupload.h"
+#include "archival/types.h"
+#include "config/configuration.h"
+#include "model/fundamental.h"
+#include "storage/disk_log_impl.h"
+#include "utils/human.h"
+
+namespace archival {
+
+static std::pair<size_t, size_t> get_low_high_segment_size(
+  size_t segment_size,
+  config::binding<std::optional<size_t>>& low_wm,
+  config::binding<std::optional<size_t>>& high_wm) {
+    auto high_watermark = high_wm().value_or(segment_size);
+    auto low_watermark = low_wm().value_or(high_watermark / 2);
+    if (low_watermark >= high_watermark) {
+        // Low watermark can't be equal to high watermark
+        // otherwise the merger want be able to find upload
+        // candidate.
+        low_watermark = high_watermark * 8 / 10;
+    }
+    return std::make_pair(low_watermark, high_watermark);
+}
+
+adjacent_segment_merger::adjacent_segment_merger(
+  ntp_archiver& parent, retry_chain_logger& ctxlog, bool is_local)
+  : _is_local(is_local)
+  , _archiver(parent)
+  , _ctxlog(ctxlog)
+  , _segment_size(config::shard_local_cfg().log_segment_size.bind())
```

Comment:
It looks like this field `_segment_size` is not used. Should it be used somewhere?